File: server/src/main/java/com/lynbrookrobotics/funkydashboard/Dataset.java
package com.lynbrookrobotics.funkydashboard;

import com.eclipsesource.json.Json;
import com.eclipsesource.json.JsonObject;
import com.eclipsesource.json.JsonValue;

public abstract class Dataset {
    String name;
    String type;

    public String getName() {
        return name;
    }

    public Dataset(String name, String type) {
        this.name = name;
        this.type = type;
    }

    public JsonValue toJSON() {
        return Json.object().add("name", name).add("type", type).add("properties", extraPropertiesJSON());
    }

    public abstract JsonValue extraPropertiesJSON();
    public abstract JsonValue currentValueJSON();
}

File: jvm/src/main/scala/com/lynbrookrobotics/funkydashboard/Dataset.scala
package com.lynbrookrobotics.funkydashboard

import play.api.libs.json.{Json, Writes}

abstract class Dataset[T](implicit valueWriter: Writes[T]) {
  val definition: DatasetDefinition

  def currentValue: T

  def handleIncomingData(data: String): Unit = {
    println(s"Unhandled incoming data: $data")
  }

  def currentValueJSON: String = {
    Json.toJson(currentValue).toString()
  }
}
